docs/rfc/0000-application-architecture.md

```diff
@@ -0,0 +1,503 @@
+
+    $XDG_RUNTIME_DIR/radicle/<srv>-<peer-id>.sock
+
+where `srv` is either "node" or "seed".
+
+The RPC API exposed follows the interface defined by `librad`'s `net::Peer`.
+
+> Note: this obviously excludes all `using_storage` operations, and is pending
+> async-ification of the `replicate` function (which then shall be promoted to a
+> `net::Peer` method).
+
+Events emitted by the node cannot be subscribed to directly over the RPC API,
+but only indirectly over the pub-sub bus. The node shall itself listen for
+pub-sub messages related to publishable updates of the repository state (eg.
+after a `git push`).
+
+> TBD: describe event types, topic names and scoping (peer id)
+
+
+### git
+
+Interactions with (logical) git repositories managed by a peer-to-peer node
+currently relies on a git remote helper and client-side refs rewriting. This is
+mainly because of a lack of library support for rewriting refs (eg. to display
+nick names alongside peer ids) on the "server". This is somewhat inconvenient,
+as the remote helper needs to be placed in the user's PATH during installation,
+and the (client) git configuration needs to be rewritten whenever the remote
+state changes. Ideally, we should be able to use a built-in git transport to
+connect to a custom git daemon, which is started on-demand via socket
+activation.
+
+Pending some advancements in the first, we can achieve this through the
+`gitoxide` and `thrussh` crates, and using the `ssh` transport for "rad
+remotes".
+
+> The obvious drawback is that this requires to bind to a TCP socket, and thus
+> breaks the desired isolation to the logged-in user. A workaround could be to
+> supply the `ProxyCommand` option to ssh, which proxies the connection over a
+> UNIX socket. While possible to achieve by modifying the user's git
+> configuration, a more robust and flexible solution might be to provide a
+> wrapper command (`rad push/pull`).
+
+The git daemon emits events to the pub-sub bus after successful pushes. Like the
+[post-receive] hook, this is one event per updated ref:
+
+```cddl
+updated = (
+    oid-old: bstr .size 20,
+    oid-new: bstr .size 20,
+    namespace: bstr,
+    ref: tstr        ; relative to namespace
+)
+```
+
+The peer id the git daemon is assuming shall be encoded in the topic name.
+
+
+### CLI
+
+The canonical `radicle-link` CLI shall follow the "subcommand" pattern familiar
+from git and other complex commandline applications. Like git, it shall be
+possible to extend the set of available subcommands by placing executables
+conforming to a naming convention (eg. `rad-<command>`) in the user's PATH. It
+SHALL NOT be possible to override builtin commands by this mechanism.
+
+Each subcommand MUST expose its functionality as a linkable library, and provide
+CBOR serialisation for its arguments and outputs.
+
+To enable [browser applications](#browser-applications), subcommands may be
+callable over the [IPC](#ipc) protocol ("CLIaaS"). The canonical CLI may provide
```

Comment:
The idea would be that it is not currently supported to have “plain” browser applications (except for things which don’t require access to the secret key). Of course, the electron `main` process could also just fork CLI commands, but calling them over a pipe is more performant, and probably more convenient that building argument lists and parsing the output. Also, if the app doesn’t require any extra functionality or command wrappers, it can be deployed without any native code.


Ultimately, it’s up to the application.

Reply:
I have a distinction between "native apps" (including electron), and "browser apps". Hope that makes it clearer.